Convert ID Integer columns to strings

robdaemon opened the pull request:
Converts any ID (PK or FK) column that's derived from a :type/Integer to
a string in query middleware (at the end of query processing.)

JavaScript treats all numbers as floats, with a 52 bit mantissa. Any
integer larger than 52 bits ends up looking like a strange floating
point number.

This will allow the UI to properly display a number n where n > 2^52 or
n < -2^52

Resolves #5816

[ci all]

Commit:
Convert ID Integer columns to strings

Converts any ID (PK or FK) column that's derived from a :type/Integer to
a string in query middleware (at the end of query processing.)

JavaScript treats all numbers as floats, with a 52 bit mantissa. Any
integer larger than 52 bits ends up looking like a strange floating
point number.

This will allow the UI to properly display a number n where n > 2^52 or
n < -2^52

Resolves #5816

[ci all]

robdaemon commented:
@paulrosenzweig @camsaul this is based on what we discussed in Slack. It has a limitation which is explained in the comment in the middleware and demonstrated in the test.

paulrosenzweig commented on src/metabase/query_processor/middleware/large_int_id.clj:
```diff
@@ -0,0 +1,47 @@
+  penalty of a comparison based on size."
+  [qp]
+  (fn [{{:keys [js-int-to-string?] :or {js-int-to-string? false}} :middleware, :as query} rff context]
+    ;; currently, this excludes `:field-literal` values like aggregations.
```
Not converting aggregated columns seem fine to me. If you sum some IDs, you probably don't have an ID anymore.

I think native queries could be an issue though. I guess fixing it for structured queries is better than not having a solution at all, but this might push us to the "full" solution more quickly.

robdaemon replied:
Yeah, native queries could be a problem. On a native query, we don't know what's an ID, we just know it's an Integer type.

camsaul commented on src/metabase/api/card.clj:
```diff
@@ -611,8 +611,9 @@
-                     :async? true)
+        query (-> (assoc (query-for-card card parameters constraints middleware)
+                         :async? true)
+                  (update :middleware #(assoc % :js-int-to-string? true)))
```
could just do `(update :middleware assoc :js-int-to-string? true)` or even `(assoc-in [:middleware :js-int-to-string?] true)`

camsaul commented on src/metabase/api/dataset.clj:
```diff
@@ -49,7 +49,9 @@
                         :card-id     source-card-id
-                        :nested?     (boolean source-card-id)}]
+                        :nested?     (boolean source-card-id)}
+        query          (update query :middleware #(-> %
```
don't need to `->` here

Commit:
code review feedback

RMMSecurity commented:
The same thing happens with UNIQUEIDENTIFIER (GUIDs) on SQLite where it breaks drill-throughs. Even basic filtering is impossible since it transforms it to hex and can't take a binary representation nor the hex shorthand it tries to use. Kind of the worst of both worlds. 

If it was just cast as a string  "X'....'" or the original GUID it would be usable. Example:

-- Original Id = {480CBDDB-8461-4EC3-9757-AA712CE079CA}

SELECT * FROM mytable WHERE id = X'480CBDDB84614EC39757AA712CE079CA'

robdaemon commented:
@RMMSecurity Thanks - can you write up an issue for that case?

RMMSecurity commented:
> @RMMSecurity Thanks - can you write up an issue for that case?

Sure thing!

sbelak commented on src/metabase/query_processor/middleware/large_int_id.clj:
```diff
@@ -0,0 +1,47 @@
+                                                       (isa? (:special_type val) :type/FK))
+                                                   (isa? (:base_type val) :type/Integer))
+                                          idx)) fields)]
+      (qp query (if js-int-to-string?
```
perhaps check here if we have any indices to update and short-circuit rather than in the reduction step?

robdaemon replied:
can do!

Commit:
skip the reduce phase if there are no columns to update

camsaul reviewed (approved):
The backend stuff looks good to me.

robdaemon commented:
going to see if I can do something about the failing cypress test.

Commit:
handle joins properly

joins have two :field-id values in them. handle them appropriately

robdaemon commented:
okay so I wasn't handling joins properly in the `(:fields)` clause of a query. Are there other conditions I'm missing, like other things in `(:fields)` that will affect the returned row?

Commit:
update cloverage build

Commit:
update tools.namespace for newer cloverage

Commit:
handle all query types more gracefully

Only in cases where we have an explicit :field-id do we try to do the
coercions to strings. It's important to not filter the :fields clause in
such a way that anything is left out - if they were, the keep-indexed
call would modify the wrong values in the result.
